### DISCUSSION TITLE
Change pins in qei.rs example

### DISCUSSION THREAD

**Discussion by franckinux** [stm32-rs/stm32f4xx-hal#757] [Q&A] [answered]
My purpose is to use both 32 bits timers (TIM2 and TIM5) as encoders. I cannot find a combination of pins that compiles.

When compiling the qei.rs example, there's no error.

`cargo build --example qei --features="tim2" --features="stm32f411"`

Now, if I change :

```
    let gpioa = dp.GPIOA.split();
    let rotary_encoder_pins = (gpioa.pa0, gpioa.pa1); 
```
to : 

    let gpioa = dp.GPIOA.split();
    let gpiob = dp.GPIOB.split();
    let rotary_encoder_pins = (gpioa.pa0, gpiob.pb3);

I get this error :

```
   Compiling stm32f4xx-hal v0.21.0 (stm32f4xx-hal)
error[E0277]: the trait bound `stm32f4xx_hal::gpio::Alternate<0>: gpio::marker::NotAlt` is not satisfied
  --> examples/qei.rs:42:57
   |
42 |     let rotary_encoder = Qei::new(rotary_encoder_timer, rotary_encoder_pins);
   |                          --------                       ^^^^^^^^^^^^^^^^^^^ the trait `gpio::marker::NotAlt` is not implemented for `stm32f4xx_hal::gpio::Alternate<0>`, which is required by `stm32f4xx_hal::gpio::Pin<'B', 3, stm32f4xx_hal::gpio::Alternate<0>>: Into<stm32f4xx_hal::gpio::alt::tim2::Ch2>`
   |                          |
   |                          required by a bound introduced by this call
   |
   = help: the following other types implement trait `gpio::marker::NotAlt`:
             DefaultMode
             stm32f4xx_hal::gpio::Output<Otype>
             stm32f4xx_hal::gpio::Analog
   = note: required for `stm32f4xx_hal::gpio::alt::tim2::Ch2` to implement `From<stm32f4xx_hal::gpio::Pin<'B', 3, stm32f4xx_hal::gpio::Alternate<0>>>`
   = note: required for `stm32f4xx_hal::gpio::Pin<'B', 3, stm32f4xx_hal::gpio::Alternate<0>>` to implement `Into<stm32f4xx_hal::gpio::alt::tim2::Ch2>`
note: required by a bound in `stm32f4xx_hal::qei::Qei::<TIM>::new`
  --> stm32f4xx-hal/src/qei.rs:45:18
   |
41 |     pub fn new(
   |            --- required by a bound in this associated function
...
45 |             impl Into<<TIM as CPin<1>>::Ch<PushPull>>,
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Qei::<TIM>::new`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `stm32f4xx-hal` (example "qei") due to 1 previous error
```
Does it mean the channels must belong to the same port ? PB3 is a legal choice for TIM2_CH2 (cf Table 9. Alternate function mapping in the reference manual).

The error messages don't help me very much.

Thanks for your attention.

**Comment by burrbull** [comment, marked as answer]
The issue with `PB3` it is already in `AlternateMode`, but different then needed. Configured for JTAG.
```rust
let rotary_encoder_pins = (gpioa.pa0, gpiob.pb3.into_alternate());
```

This is because `From` trait is not implemented for conversion between alternate modes due to conflict with implementation of `From<Self>`.
Alternative solution is to use custom trait instead of `From`.

**Reply by franckinux** [reply]
Thank you, this solves the compilation problem.

I wish I could solve it by myself but despite my attempts, I was unable to do so. This is this kind of problem that makes me say Rust is a difficult language.
